### ISSUE TITLE
consider replacing WiFi with Bluetooth LE (BLE) for battery powered devices

### ISSUE THREAD

**Issue by vladak** [vladak/shield#26]
Given the occasional problems with transferring data from the Feather V2 to the MQTT broker, it might be worthwhile exploring Bluetooth (LE) which is also likely less energy intensive. The idea is to have the Raspberry Pi run a server that would expose Bluetooth *service* with a writeable *characteristics* and the Feather would scan for it, connect and write the sensor metrics to the characteristics. The server would take the sensor values and publish them to the specified MQTT topic.

There is https://github.com/adafruit/Adafruit_CircuitPython_BLE for the client side.

There would have to be some sort of Bluetooth-to-MQTT gateway, possibly based on https://github.com/ukBaz/python-bluezero/ or perhaps even https://github.com/google/bumble/

**Comment by vladak** [comment]
The trouble is that CP 9 on Feather V2 does not currently support Bluetooth. The work has just been started on this: https://forums.adafruit.com/viewtopic.php?p=1013824

**Comment by vladak** [comment]
Due to the above I will have to put this to ice, until the ESP32 BLE support is added. Here are some links for inspiration once I get back to this:
  - CP side:
    - https://github.com/adafruit/Adafruit_CircuitPython_BLE/blob/main/examples/ble_json_central.py
    - https://github.com/adafruit/Adafruit_CircuitPython_BLE/blob/main/examples/ble_json_service.py
    - https://learn.adafruit.com/introduction-to-bluetooth-low-energy/gatt
  - Linux side:
    - https://gist.github.com/vladak/0aab7591c88554a462c96db3938b812d
    - https://stackoverflow.com/questions/48726498/python3-cant-import-dbus-even-though-its-installed
    - https://bluezero.readthedocs.io/en/stable/install_bluez.html
    - https://github.com/ukBaz/python-bluezero/blob/main/examples/cpu_temperature.py

**Comment by vladak** [comment]
This is a skeleton of the Linux side:
```python
#!/usr/bin/env python3
"""GATT Server to accept sensor values via JSON"""

import sys
sys.path.insert(0, "/usr/lib/python3/dist-packages")

import logging
import random

from bluezero import async_tools
from bluezero import adapter
from bluezero import peripheral

# Custom service uuid
SENSOR_SRVC = "51ad213f-e568-4e35-84e4-67af89c79ef0"
SENSOR_CHRC = "528ff74b-fdb8-444c-9c64-3dd5da4135ae"


def write_value(value, options):
    """
    Called when a central writes to our write characteristic.

    :param value: data sent by the central
    :param options:
    """
    # TODO: this should be JSON, publish this to MQTT
    print(f"got '{value}'")


def main(adapter_address):
    """Creation of peripheral"""
    logger = logging.getLogger('localGATT')
    logger.setLevel(logging.DEBUG)

    sensor_server = peripheral.Peripheral(adapter_address,
                                        local_name='Sensor server',
                                        appearance=1344)
    sensor_server.add_service(srv_id=1, uuid=SENSOR_SRVC, primary=True)
    sensor_server.add_characteristic(srv_id=1, chr_id=1, uuid=SENSOR_CHRC,
                                   value=[], notifying=False,
                                   flags=['write'],
                                   read_callback=None,
                                   write_callback=write_value,
                                   notify_callback=None
                                   )

    # Publish peripheral and start event loop
    sensor_server.publish()


if __name__ == '__main__':
    # Get the default adapter address and pass it to main
    # TODO: filter the adapter according to the MAC address as found using the 'hcitool dev' command.
    for adapter in list(adapter.Adapter.available()):
        print(adapter.address)

    main(adapter.address)
```

The external USB Bluetooth adapter (UB400) connected to the Raspberry Pi is `hci1` and needs to be used explicitly.

**Comment by vladak** [comment]
Eventually I used the 433 MHz radio to change the wireless transport. Although not built in, It is simpler to use code wise and has low energy requirements. Merged in 8ac60d378f50968aeaafc5340b40b3c69a0d7c72
